### PR TITLE
Support polymorphic types in typed binders

### LINKED ISSUE #2626: Support polymorphic types in typed binders
See also #2445

### PR DESCRIPTION
**Pull request by paf31**
Fixes #2626

Thanks to @LiamGoodacre for realizing that this was pretty much working already 😄

### CODE REVIEW

File: src/Language/PureScript/Parser/Declarations.hs [hunk 1]
```diff
@@ -545,7 +545,7 @@ parseBinder =
     ]
 
   -- TODO: parsePolyType when adding support for polymorphic types
```
Comment (kRITZCREEK) [inline src/Language/PureScript/Parser/Declarations.hs]: can this TODO be removed now? ;)

Comment (paf31) [inline src/Language/PureScript/Parser/Declarations.hs]: Ah, it wasn't as simple as I thought ...

**Comment by paf31** [comment]
I tried this, which I expected to fail:

```purescript
> class Default a where def :: a

> let f = \(x :: (forall a. Default a => a) -> Int) -> x def

> :type f
((forall a. (Default a) => a) -> Int) -> Int

> g \_ -> 0
0
```

I'd like to understand why this case succeeds before merging, since I would expect this to fail because we try to unify constrained types somewhere.

**Comment by rightfold** [comment]
Does this support e.g.

    data T a = A | B a

    f :: T Int -> T String
    f x@(A _ :: forall a. T a) = x
    f (B n) = B (show n)

**Comment by LiamGoodacre** [comment]
@rightfold no that's not supported

**Comment by paf31** [comment]
I've convinced myself using `trace` that this is unifying the right things. Unifying constrained types might result in errors in some cases, but for simple things like this, it should be fine.

**Pull request merged**